File: src/simulation/culture/group/centers/AspectCenter.java
package simulation.culture.group.centers;

import kotlin.Pair;
import shmp.random.RandomException;
import simulation.Event;
import simulation.culture.aspect.*;
import simulation.culture.aspect.MutableAspectPool;
import simulation.culture.group.AspectDependencyCalculator;
import simulation.space.resource.tag.ResourceTag;
import simulation.culture.aspect.dependency.*;
import simulation.space.resource.Resource;
import simulation.space.resource.tag.labeler.ResourceLabeler;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static shmp.random.RandomCollectionsKt.*;
import static shmp.random.RandomProbabilitiesKt.testProbability;
import static simulation.Controller.session;

public class AspectCenter {
    private Group group;
    private MutableAspectPool aspectPool = new MutableAspectPool(new HashSet<>());
    /**
     * Equals to aspects added on the current turn
     */
    private MutableAspectPool changedAspectPool = new MutableAspectPool(new HashSet<>());

    private List<ConverseWrapper> _converseWrappers = new ArrayList<>();
    private List<Resource> _lastResourcesForCw = new ArrayList<>();

    public AspectCenter(Group group, List<Aspect> aspects) {
        this.group = group;
        aspects.forEach(this::hardAspectAdd);
        aspectPool.getAll().forEach(a -> a.swapDependencies(this));//TODO will it swap though?
    }

    public AspectPool getAspectPool() {
        return aspectPool;
    }

    public boolean addAspect(Aspect aspect) {
        if (!aspect.isValid()) {
            return false;
        }
        if (aspectPool.contains(aspect)) {
            aspect = aspectPool.getValue(aspect.getName());
        }
        AspectDependencies _m = calculateDependencies(aspect);
        if (!aspect.isDependenciesOk(_m)) {
            return false;
        }

        addAspectNow(aspect, _m);
        return true;
    }

    private void addAspectNow(Aspect aspect, AspectDependencies dependencies) {
        Aspect _a;
        if (aspectPool.contains(aspect)) {
            _a = aspectPool.getValue(aspect);//TODO why one, add a l l
            _a.addOneDependency(dependencies);
        } else {
            _a = aspect.copy(dependencies);
            changedAspectPool.add(_a);
            if (!(_a instanceof ConverseWrapper)) {//TODO maybe should do the same in straight
                Set<Resource> allResources = new HashSet<>(group.getOverallTerritory().getDifferentResources());
                allResources.addAll(aspectPool.getProducedResources().stream()
                        .map(Pair::getFirst)
                        .collect(Collectors.toSet()));
                for (Resource resource : allResources) {
                    addConverseWrapper(_a, resource);
                }
            }
        }
    }

    void hardAspectAdd(Aspect aspect) {
        changedAspectPool.add(aspect);
        aspectPool.add(aspect);
    }

    AspectDependencies calculateDependencies(Aspect aspect) {
        AspectDependencyCalculator calculator = new AspectDependencyCalculator(
                aspectPool,
                group.getTerritoryCenter().getTerritory()
        );
        calculator.calculateDependencies(aspect);
        return calculator.getDependencies();
    }

    private void addConverseWrapper(Aspect aspect, Resource resource) { //TODO I'm adding a lot of garbage
        ConverseWrapper _w;
        if (aspect.canApplyMeaning()) {
            _w = new MeaningInserter(aspect, resource);
        } else {
            _w = new ConverseWrapper(aspect, resource);
        }
        if (!_w.isValid()) {
            return;
        }
        _converseWrappers.add(_w);
    }

    Collection<Event> mutateAspects() { //TODO separate adding of new aspects and updating old
        if (testProbability(session.rAspectAcquisition, session.random)) {
            List<Aspect> options = new ArrayList<>();
            if (session.independentCvSimpleAspectAdding) {
                if (testProbability(0.1, session.random)) {
                    options.addAll(session.world.getAspectPool().getAll());
                } else {
                    options.addAll(getAllPossibleConverseWrappers());
                }
            } else {
                options.addAll(session.world.getAspectPool().getAll());
                options.addAll(getAllPossibleConverseWrappers());
            }

            if (!options.isEmpty()) {
                Aspect _a = randomElement(options, session.random);
                if (addAspect(_a)) {
                    return Collections.singleton(new Event(
                                    Event.Type.AspectGaining,
                                    String.format("Got aspect %s by itself", _a.getName())
                            ));
                }
            }
        }
        return new ArrayList<>();
    }

    private List<ConverseWrapper> getAllPossibleConverseWrappers() {
        List<ConverseWrapper> options = new ArrayList<>(_converseWrappers); //TODO maybe do it after the middle part?
        Set<Resource> newResources = new HashSet<>(group.getOverallTerritory().getDifferentResources());
        newResources.addAll(aspectPool.getProducedResources().stream()
                .map(Pair::getFirst)
                .collect(Collectors.toSet()));
        newResources.removeAll(_lastResourcesForCw);
        for (Aspect aspect : aspectPool.filter(aspect -> !(aspect instanceof ConverseWrapper))) {
            for (Resource resource : newResources) {
                addConverseWrapper(aspect, resource);
            }
        }
        _lastResourcesForCw.addAll(newResources);
        newResources.forEach(resource -> group.getCultureCenter().getMemePool().addResourceMemes(resource));
        return options;
    }

    Set<Aspect> finishUpdate() {
        aspectPool.getAll().forEach(Aspect::finishUpdate);
        return pushAspects();
    }

    Set<Aspect> pushAspects() {
        changedAspectPool.getAll().forEach(this::addNewDependencies);
        aspectPool.addAll(changedAspectPool.getAll());
        Set<Aspect> addedAspects = changedAspectPool.getAll();
        changedAspectPool.clear();
        return addedAspects;
    }

    private void addNewDependencies(Aspect newAspect) {
        if (newAspect instanceof ConverseWrapper) {
            for (ConverseWrapper converseWrapper : aspectPool.getConverseWrappers()) {
                for (ResourceTag tag : newAspect.getTags()) {
                    if (converseWrapper.getDependencies().containsDependency(tag)) {
                        converseWrapper.getDependencies().getMap().get(tag).add(new LineDependency(
                                false,
                                converseWrapper,
                                (ConverseWrapper) newAspect
                                ));
                    }
                }
                if (newAspect.getProducedResources().stream().anyMatch(r -> converseWrapper.resource.equals(r))) {
                    converseWrapper.getDependencies().getMap().get(ResourceTag.phony()).add(new LineDependency(
                            true,
                            converseWrapper,
                            (ConverseWrapper) newAspect
                    ));
                }
            }
        }
    }

    List<Pair<Aspect, Group>> findOptions(ResourceLabeler labeler) {
        List<Pair<Aspect, Group>> options = new ArrayList<>();
        AspectLabeler aspectLabeler = new AspectLabeler(labeler);
        for (Aspect aspect : session.world.getAspectPool().getAll().stream()
                .filter(aspectLabeler::isSuitable)
                .collect(Collectors.toList())) {
            AspectDependencies _m = calculateDependencies(aspect);
            if (aspect.isDependenciesOk(_m)) {
                options.add(new Pair<>(aspect.copy(_m), null));
            }
        }

        getAllPossibleConverseWrappers().stream().filter(aspectLabeler::isSuitable)
                .forEach(wrapper -> options.add(new Pair<>(wrapper, null)));

        List<Pair<Aspect, Group>> aspects = getNeighbourAspects();
        for (Pair<Aspect, Group> pair : aspects) {
            Aspect aspect = pair.getFirst();
            Group aspectGroup = pair.getSecond();
            AspectDependencies _m = calculateDependencies(aspect);
            if (aspectLabeler.isSuitable(aspect) && aspect.isDependenciesOk(_m)) {
                aspect = aspect.copy(_m);
                options.add(new Pair<>(aspect, aspectGroup));
            }
        }
        return options;
    }

    List<Pair<Aspect, Group>> getNeighbourAspects() {
        List<Pair<Aspect, Group>> allExistingAspects = new ArrayList<>();
        for (Group neighbour : group.getRelationCenter().getRelatedGroups()) {
            allExistingAspects.addAll(neighbour.getCultureCenter().getAspectCenter().getAspectPool().getAll().stream()
                    .filter(aspect -> !(aspect instanceof ConverseWrapper)
                            || aspectPool.contains(((ConverseWrapper) aspect).aspect))
                    .map(a -> new Pair<>(a, neighbour))
                    .collect(Collectors.toList()));
        }
        return allExistingAspects;
    }

    List<Pair<Aspect, Group>> getNeighbourAspects(Predicate<Aspect> predicate) {
        return getNeighbourAspects().stream()
                .filter(p -> predicate.test(p.getFirst()))
                .collect(Collectors.toList());
    }

    Collection<Event> adoptAspects(Group group) {
        if (!session.isTime(session.groupTurnsBetweenAdopts)) {
            return new ArrayList<>();
        }
        List<Pair<Aspect, Group>> allExistingAspects = getNeighbourAspects(a -> !changedAspectPool.contains(a));

        if (!allExistingAspects.isEmpty()) {
            try {
                Pair<Aspect, Group> pair = randomElement(
                        allExistingAspects,
                        p -> p.getFirst().getUsefulness() * group.getRelationCenter().getNormalizedRelation(p.getSecond()),
                        session.random
                );
                if (addAspect(pair.getFirst())) {
                    return Collections.singleton(new Event(
                                    Event.Type.AspectGaining,
                                    String.format(
                                            "Got aspect %s from group %s",
                                            pair.getFirst().getName(), pair.getSecond().name)
                            ));
                }
            } catch (Exception e) {
                if (e instanceof RandomException) {
                    int i = 0;//TODO
                }
            }
        }
        return  new ArrayList<>();
    }
}

File: src/simulation/culture/group/centers/AspectCenter.kt
package simulation.culture.group.centers

import shmp.random.RandomException
import shmp.random.randomElement
import shmp.random.testProbability
import simulation.Controller.*
import simulation.Event
import simulation.culture.aspect.*
import simulation.culture.aspect.dependency.AspectDependencies
import simulation.culture.aspect.dependency.LineDependency
import simulation.culture.aspect.labeler.ProducedLabeler
import simulation.culture.group.AspectDependencyCalculator
import simulation.culture.group.GroupError
import simulation.culture.group.convert
import simulation.space.resource.Resource
import simulation.space.resource.tag.ResourceTag
import simulation.space.resource.tag.labeler.ResourceLabeler
import java.util.*
import java.util.function.Predicate

class AspectCenter(private val group: Group, aspects: List<Aspect>) {
    private val _mutableAspectPool = MutableAspectPool(HashSet())


    /**
     * Equals to aspects added on the current turn
     */
    private val changedAspectPool = MutableAspectPool(HashSet())
    private val _converseWrappers = mutableSetOf<ConverseWrapper>()
    private val _lastResourcesForCw: MutableList<Resource> = ArrayList()

    init {
        aspects.forEach { hardAspectAdd(it) }
        _mutableAspectPool.all.forEach { it.swapDependencies(this) } //TODO will it swap though?
    }

    val aspectPool: AspectPool
        get() = _mutableAspectPool

    fun addAspect(aspect: Aspect): Boolean {
        var currentAspect = aspect
        if (!currentAspect.isValid)
            return false
        if (_mutableAspectPool.contains(currentAspect))
            currentAspect = _mutableAspectPool.getValue(currentAspect.name)
        val dependencies = calculateDependencies(currentAspect)
        if (!currentAspect.isDependenciesOk(dependencies))
            return false
        addAspectNow(currentAspect, dependencies)
        return true
    }

    private fun addAspectNow(aspect: Aspect, dependencies: AspectDependencies) {
        val currentAspect: Aspect
        if (_mutableAspectPool.contains(aspect)) {
            currentAspect = _mutableAspectPool.getValue(aspect) //TODO why one, add a l l
            currentAspect.addOneDependency(dependencies)
        } else {
            currentAspect = aspect.copy(dependencies)
            changedAspectPool.add(currentAspect)
            if (currentAspect !is ConverseWrapper) { //TODO maybe should do the same in straight
                val allResources: MutableSet<Resource> = HashSet(group.overallTerritory.differentResources)
                allResources.addAll(_mutableAspectPool.producedResources.map { it.first })
                for (resource in allResources)
                    addConverseWrapper(currentAspect, resource)
            }
        }
    }

    private fun hardAspectAdd(aspect: Aspect) {
        changedAspectPool.add(aspect)
        _mutableAspectPool.add(aspect)
    }

    private fun calculateDependencies(aspect: Aspect): AspectDependencies {
        val calculator = AspectDependencyCalculator(
                _mutableAspectPool,
                group.territoryCenter.territory
        )
        calculator.calculateDependencies(aspect)
        return calculator.dependencies
    }

    private fun addConverseWrapper(aspect: Aspect, resource: Resource) { //TODO I'm adding a lot of garbage
        val wrapper = (if (aspect.canApplyMeaning())
            MeaningInserter(aspect, resource)
        else
            ConverseWrapper(aspect, resource))
        if (!wrapper.isValid)
            return
        _converseWrappers.add(wrapper)
    }

    fun mutateAspects(): Collection<Event> { //TODO separate adding of new aspects and updating old
        if (testProbability(session.rAspectAcquisition, session.random)) {
            val options: MutableList<Aspect> = ArrayList()
            if (session.independentCvSimpleAspectAdding) {
                if (testProbability(0.1, session.random))
                    options.addAll(session.world.aspectPool.all)
                else
                    options.addAll(allPossibleConverseWrappers)
            } else {
                options.addAll(session.world.aspectPool.all)
                options.addAll(allPossibleConverseWrappers)
            }
            if (options.isNotEmpty()) {
                val aspect = randomElement(options, session.random)
                if (aspect is ConverseWrapper && !aspectPool.contains(aspect.aspect)) {
                    return listOf()
                }
                if (addAspect(aspect)) {
                    return setOf(Event(
                            Event.Type.AspectGaining, String.format("Got aspect %s by itself", aspect.name)
                    ))
                }
            }
        }
        return listOf()
    }

    private val allPossibleConverseWrappers: List<ConverseWrapper>
        get() {
            val options: List<ConverseWrapper> = ArrayList(_converseWrappers) //TODO maybe do it after the middle part?
            val newResources: MutableSet<Resource> = HashSet(group.overallTerritory.differentResources)
            newResources.addAll(_mutableAspectPool.producedResources.map { it.first })
            newResources.removeAll(_lastResourcesForCw)
            for (aspect in _mutableAspectPool.filter { it !is ConverseWrapper })
                for (resource in newResources)
                    addConverseWrapper(aspect, resource)
            _lastResourcesForCw.addAll(newResources)
            newResources.forEach { group.cultureCenter.memePool.addResourceMemes(it) }
            return options
        }

    fun finishUpdate(): Set<Aspect> {
        _mutableAspectPool.all.forEach { it.finishUpdate() }
        return pushAspects()
    }

    fun pushAspects(): Set<Aspect> {
        changedAspectPool.all.forEach { addNewDependencies(it) }
        _mutableAspectPool.addAll(changedAspectPool.all)
        val addedAspects = changedAspectPool.all
        changedAspectPool.clear()
        return addedAspects
    }

    private fun addNewDependencies(newAspect: Aspect) {
        if (newAspect is ConverseWrapper)
            for (converseWrapper in _mutableAspectPool.converseWrappers) {
                for (tag in newAspect.tags)
                    if (converseWrapper.dependencies.containsDependency(tag))
                        converseWrapper.dependencies.map[tag]!!.add(LineDependency(
                                false,
                                converseWrapper,
                                newAspect
                        ))
                if (newAspect.producedResources.any { converseWrapper.resource == it })
                    converseWrapper.dependencies.map[ResourceTag.phony()]!!.add(LineDependency(
                            true,
                            converseWrapper,
                            newAspect
                    ))
            }
    }

    fun findOptions(labeler: ResourceLabeler): List<Pair<Aspect, Group?>> {
        val options: MutableList<Pair<Aspect, Group?>> = ArrayList()
        val aspectLabeler = ProducedLabeler(labeler)
        for (aspect in session.world.aspectPool.all.filter { aspectLabeler.isSuitable(it) }) {
            val dependencies = calculateDependencies(aspect)
            if (aspect.isDependenciesOk(dependencies))
                options.add(Pair<Aspect, Group?>(aspect.copy(dependencies), null))
        }
        allPossibleConverseWrappers
                .filter { aspectLabeler.isSuitable(it) }
                .forEach { options.add(Pair<Aspect, Group?>(it, null)) }
        val aspects = convert(newNeighbourAspects)
        for (box in aspects) {
            var aspect = box.aspect
            val aspectGroup = box.group
            if (aspectLabeler.isSuitable(aspect)) {
                val dependencies = calculateDependencies(aspect)
                if (aspect.isDependenciesOk(dependencies)) {
                    aspect = aspect.copy(dependencies)
                    options.add(Pair(aspect, aspectGroup))
                }
            }
        }
        return options
    }

    private val neighbourAspects: List<Pair<Aspect, Group>>
        get() {
            val allExistingAspects: MutableList<Pair<Aspect, Group>> = ArrayList()
            for (neighbour in group.relationCenter.relatedGroups) {
                allExistingAspects.addAll(
                        neighbour.cultureCenter.aspectCenter._mutableAspectPool.all
                                .filter { (it !is ConverseWrapper || _mutableAspectPool.contains(it.aspect)) }
                                .map { Pair(it, neighbour) }
                )
            }
            return allExistingAspects
        }

    private val newNeighbourAspects: List<Pair<Aspect, Group>>
        get() = neighbourAspects.filter { (a) -> !_mutableAspectPool.contains(a) }

    private fun getNeighbourAspects(predicate: Predicate<Aspect>) =
            neighbourAspects.filter { (a) -> predicate.test(a) }

    fun adoptAspects(group: Group): Collection<Event> {
        if (!session.isTime(session.groupTurnsBetweenAdopts))
            return ArrayList()
        val allExistingAspects = getNeighbourAspects(Predicate { !changedAspectPool.contains(it) })
        if (allExistingAspects.isNotEmpty()) try {
            val (aspect, aspectGroup) = randomElement(
                    allExistingAspects,
                    { (a, g) -> a.usefulness * group.relationCenter.getNormalizedRelation(g) },
                    session.random
            )
            if (addAspect(aspect)) return setOf(Event(
                    Event.Type.AspectGaining, String.format(
                    "Got aspect %s from group %s",
                    aspect.name, aspectGroup.name)))
        } catch (e: Exception) {
            if (e is RandomException) {
                val i = 0 //TODO
            }
        }
        return ArrayList()
    }

    fun update(crucialAspects: Collection<Aspect>) {
        val unimportantAspects = aspectPool.all
                .filter { it.usefulness < session.aspectFalloff }
                .filter { it !in crucialAspects }
        if (unimportantAspects.isEmpty()) return
        val aspect = randomElement(unimportantAspects, session.random)
        if (aspect !in aspectPool.converseWrappers.map { it.aspect }) {
            if (_mutableAspectPool.remove(aspect) && aspect !is ConverseWrapper) {
                _converseWrappers.removeIf { it.aspect == aspect }
            }
        }
    }
}